### PR TITLE
[JENKINS-57894] Added data bound setters

### PR DESCRIPTION
**Pull request by baymac**
As discussed with @casz, moved some parameters to DataBoundSetter from DataBoundConstructor.

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/gitlabserver/servers/GitLabServer.java [hunk 1]
```diff
@@ -140,23 +141,46 @@ public String getCredentialsId() {
...
      * @param credentialsId The {@link StandardUsernamePasswordCredentials#getId()} of the credentials to use for
-     *                      auto-management of hooks.
+     *                      GitLab Server Authentication to access GitLab APIs
      */
     @DataBoundConstructor
-    public GitLabServer(@Nonnull String name, @NonNull String serverUrl, boolean manageHooks,
-                        @CheckForNull String credentialsId) {
-        this.manageHooks = manageHooks;
+    public GitLabServer(@CheckForNull String credentialsId) {
         this.credentialsId = credentialsId;
```
Comment (jetersen) [inline src/main/java/io/jenkins/plugins/gitlabserver/servers/GitLabServer.java]: I don't think credentials id is mandatory if the repo is public.

Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabserver/servers/GitLabServer.java]: Okay. Just wondering if GitHub Plugin did it wrongly?

**Comment by baymac** [comment]
Just a quick question. Is it okay to not keep any instance variables `final` in `GitLabServer`?

**Comment by jetersen** [comment]
it's fine not keep any final but I'd advise that serverUrl is final so it cannot be changed after the object is created.

**Comment by baymac** [comment]
I am wondering how that works. Because I am able to change the `serverUrl` even when it is declared `final`.

**Comment by jetersen** [comment]
You cannot programmatically change it :sweat_smile:
The object gets created when serialized from JSON to Java on `configure` method and then saved to XML eventually.

**Comment by jetersen** [comment]
When it is marked `final`, you cannot use a "setServerUrl" to change the value nor can you inside the class do `this.serverUrl = "HAHA"` anywhere but inside the constructor and you can only do it once.

You will be met with: `Variable 'serverUrl' might already have been assigned to`

**Comment by baymac** [comment]
Can you explain little further? My understanding is `GitLabServer` object is created when I add a `GitLab Server` from the dropdown in Global Config. Since the `serverUrl` field has default value, the DataBoundConstructor is called once and `serverUrl` is set for the `GitLabServer` object. So how am I able to change a final `serverUrl` for the same object?

**Comment by jetersen** [comment]
When you click `add` inside the global config actually NOTHING has happened yet (a truth with modification, stapler does some POJO inspection and presents the config.jelly for `GitLabServer`) :sweat: That is pure UI.

It is only when you click "save" that everything is sent as a request to Jenkins as JSON and then serialized inside each `configure` method for each JSON object.

`Go To` in IntelliJ is your friend, right click `bindJSONToList` inside `configure` method, go into the submenu `Go To`, choose `Implementation(s)` :sweat_smile: If you want to understand what Stapler does behind the scene.

You can also attach the debugger and start exploring what it called when. :smile:

**Comment by baymac** [comment]
Thanks, will do that later today.

**Comment by jetersen** [comment]
I went nuclear on the broken test so I could rebase my PR https://github.com/baymac/gitlab-branch-source-plugin/commit/13eb5706f5ef0ddfcb850754aed0be77a8ed5663

**Pull request merged**